### DISCUSSION TITLE
Option to change the default java version.

### DISCUSSION THREAD

**Discussion by jmccay-work** [apache/netbeans#5488] [Ideas]
It would be great if you could change the default version.  My Linux system uses a default version of OpenJDK 11.  NetBeans 16 has the java version set to 17.  Every time I open a project for the first time, it throws an error.  The Gradle version I am using is 6.8.2 which doesn't support 17.   Once I set the version of the JDK to 11, it is fine.  I actually don't use 17 at all.

**Comment by mbien** [comment]
the default java platform is currently the same as the JDK NetBeans is running on. This makes sense if you think about it because this is the only JDK installation NB knows for sure that it exists.

It would be nice to be able to change this in the UI, but it is also somewhat risky, it can be changed in the config file though.
https://github.com/apache/netbeans#log-config-and-cache-locations
see `netbeans_jdkhome` property.

If the path is set wrong and NB doesn't start the user knows at least already where to fix it :)

there are also community bundles available which often ship with a very recent JDK all in one package, they are listed on the download page under the main downloads - this problem doesn't appear there either since the default JDK would be the most recent one already.

**Reply by neilcsmith-net** [reply]
The Maven support actually allows for an alternative platform for all Maven projects, although the UI there is a bit odd (project properties UI shows the wrong default).  Could extend this to other build systems?  Or a general project default JDK?

The community installers currently bundle JDK 17 rather than JDK 19, which might get reviewed again in future (mainly triggered by Gradle concerns).

**Comment by jmccay-work** [comment]
Why don't you treat them as 2 different properties?  There's the JDK that NB uses to run, and there is the one NB uses to compile.  NB can always be linked to the one it need to run, and changing the default JDK version would only apply to the JDK used when opening and compiling projects.

**Reply by mbien** [reply]
but this also assumes that all projects you open use the same JDK - so I am not sure how useful this actually would be. I personally don't really do that since different projects usually require different JDKs.

**Comment by jmccay-work** [comment]
Yes, you are correct, but when opening an existing Gradle project for the first time,  I am constantly getting an annoying error message saying the project can't be loaded.  I have to do 2 things.  First, I set the project JDK version from Java 17 to Java 11.  Then I perform a first build on the project.  These 2 things fix the issues, and the project loads without any problems.  It's just annoying to have to do this every time I open a new project.  I am pushing the envelop at my company by using Java 11, but this project I am working on doesn't interface directly with the current project so I can get away with not using Java 8.

**Comment by neilcsmith-net** [comment]
I can definitely see the use for this.  I use project groups a lot, and being able to set this based on active project groups would be really useful.  I had not realised that project groups have an API for additional properties.  You can set up a custom Maven home for a project group (learn something new every day!).  Being able to configure a Java platform for all projects in a group, across build system, would be really useful.

For info - https://github.com/apache/netbeans/blob/master/java/maven/src/org/netbeans/modules/maven/options/MavenGroupCategory.java

**Reply by mbien** [reply]
the use case would be to use project groups to switch the JDKs of multiple projects at the same time instead of having to do it one by one?

**Reply by neilcsmith-net** [reply]
Yes.  But potentially stored in the group config rather than the project itself.  Need to look how that configuration works now.  Something to explore in a quieter time! :smile:  But I can certainly see a benefit to group scopes, or profiles, for configuration options.

**Reply by mbien** [reply]
I like this idea
